### PR TITLE
feat: display an upgrade message for the CLI

### LINKED ISSUE #930: feature: display an upgrade message for the CLI
Tailcall CLI needs to display a message if an upgrade is available. Here is an example of how github cli does this:

Screenshot 2024-01-13 at 11 50 24 AM

**Technical Requirements:**
- Make sure it doesn't make the overall experience slower.
- Suggest upgrade methods should be based on how it was installed originally.

### PR DESCRIPTION
**Pull request by neo773**
This PR displays an upgrade message for the CLI if there's an update available

image

**Issue Reference(s):**  
Fixes #930

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

/claim #930

### CODE REVIEW

**Comment by neo773** [comment]
@tusharmath 
Could I get a review on this?

**Comment by tusharmath** [comment]
This is a lot of code to maintain. Is there a crate for this? Posted this on reddit to get some help - https://www.reddit.com/r/rust/comments/196f0kr/crate_for_checking_if_a_newer_version_of_the/

**Comment by neo773** [comment]
@tusharmath Could I get a re-review on this?
I refactored the code use update-informer crate

File: src/cli/update_checker.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+use which::which;
+
+enum InstallationMethod {
+  Npm,
+  Brew,
+  Direct,
+}
+
+fn get_installation_method() -> InstallationMethod {
+  let output = std::process::Command::new("npm").arg("ls").arg("--global").output();
```
Comment (meskill) [inline src/cli/update_checker.rs:12]: use could actually install it locally to project and use it with `npx tailcall`.

I think we could validate some available env variables that set by npm and other package manager to figure out how program is ran.

File: src/cli/update_checker.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  if let Ok(output) = output {
+    let output_str = String::from_utf8_lossy(&output.stdout);
+    if output_str.contains("@tailcallhq/tailcall") {
+      return InstallationMethod::Npm;
+    }
+  }
+
+  if let Ok(result) = which("tailcall") {
+    if result.to_str().unwrap().contains("homebrew") {
```
Comment (meskill) [inline src/cli/update_checker.rs:22]: to_str could fail actually. Consider to not create panics

File: src/cli/update_checker.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+  InstallationMethod::Direct
+}
+
+pub async fn check_for_update() {
+  let name: &str = "tailcallhq/tailcall";
+  let current_version: &str = match option_env!("APP_VERSION") {
+    Some(version) => version,
+    _ => return,
+  };
```
Comment (meskill) [inline src/cli/update_checker.rs:35]:
```suggestion
  let Some(current_version) = option_env!("APP_VERSION") else {
      return;
  };
```

**Review by meskill — CHANGES_REQUESTED** [review]

**Comment by neo773** [comment]
@meskill 
Pushed a new commit

File: src/cli/tc.rs [hunk 4]
```diff
@@ -21,8 +23,9 @@ const YML_FILE_NAME: &str = ".graphqlrc.yml";
 
 pub async fn run() -> anyhow::Result<()> {
   let cli = Cli::parse();
-
   logger_init();
+  let rt = Runtime::new().unwrap();
+  rt.spawn(update_checker::check_for_update());
```
Comment (tusharmath) [inline src/cli/tc.rs:28]: I don't think we should need this.

**Comment by tusharmath** [comment]
Getting the following error.

```
❯ APP_VERSION=0.5.0 cargo run -- check ./examples/jsonplaceholder.graphql
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /Users/tushar/Documents/Projects/tailcall/cloudflare/Cargo.toml
workspace: /Users/tushar/Documents/Projects/tailcall/Cargo.toml
    Finished dev [unoptimized + debuginfo] target(s) in 0.40s
     Running `target/debug/tailcall check ./examples/jsonplaceholder.graphql`
[2024-01-27T06:33:45Z INFO  tailcall::cli::file] File read: ./examples/jsonplaceholder.graphql ... ok
[2024-01-27T06:33:45Z INFO  tailcall::cli::tc] Config successfully validated
N + 1: 1

thread 'main' panicked at /Users/tushar/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/blocking/shutdown.rs:51:21:
Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

**Comment by neo773** [comment]
@tusharmath 
Fixed
4B1D2D1D-7CB1-498C-9DBB-8FC1A7BA180C

**Review by tusharmath — APPROVED** [review]

**Review by meskill — APPROVED** [review]

**Pull request merged**